### PR TITLE
fix(gossipsub): prevent erroneously duplicate message IDs

### LINKED ISSUE #3714: Duplicated error with gossipsub messages
## Summary
Since 0.51.2, gossipsub messages would show duplicated message error when publishing. This is likely due to the changes brought about in https://github.com/libp2p/rust-libp2p/pull/3551, where after the initial message being publish that sequential messages of any kind (the same or random) would produce this error. This seems to change after a specific amount of time. This is using the default `Config` without any changes to the `message_id_fn` configuration (can be reproduced with the chat example by removing the `message_id_fn`). 

## Expected behaviour

Be able to publish messages in a topic

## Actual behaviour
Sequential messages after the initial publish returns a error

<details><summary>(Partial) Debug Output</summary>
<p>

```
[2023-03-30T21:40:25Z DEBUG libp2p_gossipsub::behaviour] Starting heartbeat
[2023-03-30T21:40:25Z DEBUG libp2p_gossipsub::behaviour] HEARTBEAT: Mesh low. Topic: test-net Contains: 0 needs: 5
[2023-03-30T21:40:25Z DEBUG libp2p_gossipsub::behaviour] RANDOM PEERS: Got 0 peers
[2023-03-30T21:40:25Z DEBUG libp2p_gossipsub::behaviour] Updating mesh, new mesh: {}
[2023-03-30T21:40:25Z DEBUG libp2p_gossipsub::behaviour] Completed Heartbeat
hello world
12D3KooWJoGWs5G6z3yoSjenNDgmxu3fSo9KS4PyrLHvTn8NjNLq15547085688055217085
[2023-03-30T21:40:32Z DEBUG libp2p_gossipsub::behaviour] Published message: MessageId(313244334b6f6f574a6f4757733547367a33796f536a656e4e44676d78753366536f394b53345079724c4876546e384e6a4e4c713135353437303835363838303535323137303835)
[2023-03-30T21:40:32Z DEBUG libp2p_gossipsub::behaviour] New peer type found: Gossipsub v1.1 for peer: 12D3KooWGebADVb2Z64fCEmCgzxjqEAL42wvRpdmdzksT4UVRNGQ
world hello
12D3KooWJoGWs5G6z3yoSjenNDgmxu3fSo9KS4PyrLHvTn8NjNLq15547085688055217085
[2023-03-30T21:40:34Z WARN  libp2p_gossipsub::behaviour] Not publishing a message that has already been published. Msg-id 313244334b6f6f574a6f4757733547367a33796f536a656e4e44676d78753366536f394b53345079724c4876546e384e6a4e4c713135353437303835363838303535323137303835
Publish error: Duplicate
[2023-03-30T21:40:35Z DEBUG libp2p_gossipsub::behaviour] Starting heartbeat
[2023-03-30T21:40:35Z DEBUG libp2p_gossipsub::behaviour] HEARTBEAT: Mesh low. Topic: test-net Contains: 0 needs: 5
[2023-03-30T21:40:35Z DEBUG libp2p_gossipsub::behaviour] RANDOM PEERS: Got 0 peers
[2023-03-30T21:40:35Z DEBUG libp2p_gossipsub::behaviour] Updating mesh, new mesh: {}
[2023-03-30T21:40:35Z DEBUG libp2p_gossipsub::behaviour] RANDOM PEERS: Got 0 peers
[2023-03-30T21:40:35Z DEBUG libp2p_gossipsub::behaviour] Gossiping IHAVE to 0 peers.
[2023-03-30T21:40:35Z DEBUG libp2p_gossipsub::behaviour] Completed Heartbeat
```
</p>
</details>

## Version

- libp2p version (version number, commit, or branch): 0.51.2

### PR DESCRIPTION
**Pull request by dariusc93**
## Description

Previously, we only mutably borrowed the `last_seq_no` in the current scope but did not modify the underlying number. This is because `u64` is copy and calling `wrapping_add` consumes `self` so the compiler just copied it. We introduce a new-type instead that is not `Copy`.

Additionally, `wrapping_add` and initializing with a random u64 might actually warp the number and thus not give us sequential numbers as intended in #3551. To solve this, we initialize with the current unix timestamp in nanoseconds. This allows a node to publish 1000000 messages a second and still not reuse sequence numbers even after a restart / re-initialization of the configuration. This is also what the go implementation does.

Resolves #3714.

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>
  

## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/gossipsub/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,8 @@
+## 0.44.3 - unreleased
+- Mutablely borrow last sequential number. See [PR 3716]
```
Comment (thomaseizinger) [inline protocols/gossipsub/CHANGELOG.md]:
```suggestion

- Fix erroneously duplicate message IDs. See [PR 3716].
```

File: protocols/gossipsub/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,8 @@
+## 0.44.3 - unreleased
```
Comment (thomaseizinger) [inline protocols/gossipsub/CHANGELOG.md:1]: Can you bump the version in the Cargo.toml too please?

**Review by thomaseizinger — APPROVED** [review]
I updated the title and PR description, let me know if I understood the problem and fix correctly!

**Comment by thomaseizinger** [comment]
cc @divagant-martian In case you are also upgrading to 0.51.2 already.

File: protocols/gossipsub/CHANGELOG.md (hunk 2 above)
Comment (dariusc93) [reply protocols/gossipsub/CHANGELOG.md:1]: Done

**Comment by thomaseizinger** [comment]
Lockfile needs updating!

**Comment by thomaseizinger** [comment]
I've pushed an improved fix that should make handling this slightly safer.

cc @AgeManning I've also changed this to start with the current time. The thinking was that if we want to be strictly increasing, `wrapping_add` is the wrong function as we might actually **not** increase the sequence number if we start with a really high random number. Instead, I am using milliseconds of the current unix timestamp. This means, even if a node would start and burst out messages, then shut down and start again, they have to send more than 1000 messages per second before they'd get into a state where they'd send out messages that other nodes perceive as duplicates. I think that is reasonably safe.

I also chose to `expect` the wrapping. Even if we send 1 million messages a second, we'd only run out of space after 584942 years if my [calculation](https://duckduckgo.com/?q=(2%5E64+-+1680533430000)+%2F+1000000+%2F+60+%2F+60+%2F+24+%2F+365&t=ffab&ia=calculator) is correct.

**Comment by thomaseizinger** [comment]
cc @MarcoPolo This PR changes our implementation to be the exact same as go.

**Review by AgeManning — APPROVED** [review]
Yep this makes sense to me, the panic at u64::max() is for someone living 600k years in the future's problem (if they can keep a libp2p node running that long without updating :).) 

This also doesn't effect us as we don't use sequence numbers, so all good from our end, thanks for the ping :).

**Comment by thomaseizinger** [comment]
Thank you @dariusc93 for the initial discovery of the bug and the pull request.

**Review by mxinden — APPROVED** [review]

**Comment by mxinden** [comment]
Thanks @dariusc93 for debugging and proposing a patch!

---

@thomaseizinger shall I cut a new patch release already, or do you want to wait for https://github.com/libp2p/rust-libp2p/pull/3625?

**Pull request merged**